pkg/applicationserver/io/pubsub/pubsub.go

```diff
@@ -235,27 +237,31 @@ func (ps *PubSub) start(ctx context.Context, pb *ttnpb.ApplicationPubSub) (err e
 	defer func() {
 		cancel(err)
 	}()
+
 	i := &integration{
 		ApplicationPubSub: *pb,
 		ctx:               ctx,
 		cancel:            cancel,
 		closed:            make(chan struct{}),
 		server:            ps.server,
 	}
+	defer close(i.closed)
 	if _, loaded := ps.integrations.LoadOrStore(psUID, i); loaded {
-		log.FromContext(ctx).Debug("Integration already started")
+		logger.Debug("Pub/sub already started")
 		return errAlreadyConfigured.WithAttributes("application_uid", appUID, "pub_sub_id", pb.PubSubID)
 	}
-	go func() {
-		<-ctx.Done()
-		ps.integrationErrors.Store(psUID, ctx.Err())
-		ps.integrations.Delete(psUID)
-		if err := ctx.Err(); err != nil && !errors.IsCanceled(err) {
-			log.FromContext(ctx).WithError(err).Warn("Integration failed")
+	defer ps.integrations.Delete(psUID)
+
+	defer func() {
+		if err != nil && !errors.IsCanceled(err) {
+			logger.WithError(err).Warn("Pub/sub failed")
 			registerIntegrationFail(ctx, i, err)
+		} else {
+			logger.Info("Pub/sub canceled")
+			registerIntegrationStop(ctx, i)
 		}
-		close(i.closed)
```

Comment:
Closing `i.closed` here has the same effect as the above mentioned deletion - allows a new integration to start too early.